### PR TITLE
vim-patch:8.1.{1734,1742,1743}: highlight, match and screen.c refactoring

### PR DESCRIPTION
**Pull request by lewis6991**
```
    vim-patch:8.1.1734: the evalfunc.c file is too big

    Problem:    The evalfunc.c file is too big.
    Solution:   Move some functions to other files.
```
https://github.com/vim/vim/commit/29b7d7a9aac591f920edb89241c8cde27378e50b

```
    vim-patch:8.1.1742: still some match functions in evalfunc.c

    Problem:    Still some match functions in evalfunc.c.
    Solution:   Move them to highlight.c.
```

https://github.com/vim/vim/commit/7dfb016d25e3e3e1f4411026dda21d1536f21acc

```
    vim-patch:8.1.1743: 'hlsearch' and match highlighting in the wrong place

    Problem:  'hlsearch' and match highlighting in the wrong place.
    Solution: Move highlighting from inside screen functions to highlight.c.
```
https://github.com/vim/vim/commit/bbca7732e8a3deb6e5dcf84739579a2667a75475

### CODE REVIEW

**Review by seandewar — APPROVED** [review]
LGTM I think. Didn't see any (relevant) sneaky undocumented changes made in the patch.
Might also be worth porting 8.1.1742 at some point if it's possible

File: src/nvim/screen.c [hunk 1]
```diff
@@ -2599,66 +2591,12 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
...
-    }
-    if (shl != &search_hl && cur != NULL) {
-      cur = cur->next;
-    }
+  if (!number_only) {
+    v = (long)(ptr - line);
+    area_highlighting |= prepare_search_hl_line(wp, lnum, (colnr_T)v,
+                                                &line, &search_hl, &search_attr,
+                                                &search_attr_from_match, has_fold, end_fill);
+    ptr = line + v; // "line" may have been updated
```
Comment (zeertzjq) [inline src/nvim/screen.c:2599]:
```suggestion
    ptr = line + v;  // "line" may have been updated
```

**Comment by lewis6991** [comment]
> Might also be worth porting 8.1.1742 at some point if it's possible

Done

**Comment by zeertzjq** [comment]
a lot of crashes in CI

**Comment by lewis6991** [comment]
Yep, debugging.

**Comment by lewis6991** [comment]
Was missing a couple of pointer derefs. Should be good now.

**Comment by zeertzjq** [comment]
`match_add()` and `match_delete()` can be made static now.

**Comment by lewis6991** [comment]
I'm going to be moving those functions again with 8.2.1078 after this so I don't think it's worth it.

File: src/nvim/highlight_group.c [hunk 2]
```diff
@@ -3077,6 +3081,508 @@ matchitem_T *get_match(win_T *wp, int id)
...
+
+/// Prepare for 'hlsearch' and match highlighting in one window line.
+/// Return TRUE if there is such highlighting and set "search_attr" to the
+/// current highlight attribute.
+int prepare_search_hl_line(win_T *wp, linenr_T lnum, colnr_T mincol, char_u **line,
+                           match_T *search_hl, int *search_attr, bool *search_attr_from_match,
+                           bool has_fold, bool end_fill)
+{
+  // Handle highlighting the last used search pattern and matches.
+  // Do this for both search_hl and the match list.
```
Comment (zeertzjq) [inline src/nvim/highlight_group.c:3356]: I feel this comment looks a bit strange. Maybe it is better to put this above the `while`.

File: src/nvim/screen.c [hunk 3]
```diff
@@ -2599,66 +2591,12 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
...
-      area_highlighting = true;
-    }
-    if (shl != &search_hl && cur != NULL) {
-      cur = cur->next;
-    }
+  if (!number_only) {
+    v = (long)(ptr - line);
+    area_highlighting |= prepare_search_hl_line(wp, lnum, (colnr_T)v,
+                                                &line, &search_hl, &search_attr,
+                                                &search_attr_from_match, has_fold, end_fill);
```
Comment (zeertzjq) [inline src/nvim/screen.c:2598]: Maybe it is better to skip this function call if either of `has_fold` or `end_fill` is true.

File: src/nvim/highlight_group.c [hunk 4]
```diff
@@ -2797,3 +2807,1164 @@ int name_to_ctermcolor(const char *name)
...
+  return false;
+}
+
+/// Get highlighting for the char after the text in "char_attr" from 'hlsearch'
+/// or match highlighting.
+void get_search_match_hl(win_T *wp, match_T *search_hl, long col, int *char_attr)
+{
+  matchitem_T *cur = wp->w_match_head;  // points to the match list
+  match_T     *shl;                     // points to search_hl or a match
+  int          shl_flag = false;        // flag to indicate whether search_hl
```
Comment (seandewar) [inline src/nvim/highlight_group.c:3564]: bool

**Comment by lewis6991** [comment]
Thanks for the reviews. Think I got all the comments.

**Review by zeertzjq — APPROVED** [review]

**Comment by zeertzjq** [comment]
I think it is better to have a rebase now.

**Review by seandewar — APPROVED** [review]

**Pull request merged**